Design/3093-graphql-defer.md

```diff
@@ -0,0 +1,356 @@
+      dataHandler: dataHandler,
+      errorHandler: errorHandler
+    )
+  }
+}
+
+// Sample protocol for multipart specification parsing
+protocol MultipartResponseSpecificationParser {
+  static var protocolSpec: String { get }
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  )
+}
+
+// Sample implementations of multipart specification parsers
+
+struct SubscriptionResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "subscriptionSpec=1.0"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // parsing code currently in MultipartResponseParsingInterceptor
+  }
+}
+
+struct DeferResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "deferSpec=20220824"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // new code to parse the defer specification
+  }
+}
+```
+
+#### Response data
+
+The initial response data and data received in each incremental response will need to be retained and combined so that each incremental response can insert the latest received incremental response data at the correct path and return an up-to-date response to the request callback.
+
+The data being retained and combined should not require another pass through the GraphQL executor though.
+
+### Completion handler
+
+`GraphQLResult` should be modified to provide query completion blocks with a high-level abstraction of whether the request has been fulfilled or is still in progress. This prevents clients from having to dig into the deferred fragments to identify the state of the overall request.
+
+Potential solutions:
+1. Introduce a new property on the `GraphQLResult` type that can be used to express the state of the request.
+
+```swift
+// New Response type and property
+public struct GraphQLResult<Data: RootSelectionSet> {
+  // other properties and types not shown
+
+  public enum Response {
+    case partial
+    case complete
+  }
+
+  public let response: Response
+}
+
+// Sample usage in an app completion block
+client.fetch(query: ExampleQuery()) { result in
+  switch (result) {
+  case let .success(data):
+    switch (data.response) {
+    case .complete:
+    case .partial:
+    }
+  case let .failure(error):
+  }
+}
+```
+
+2. Another way which may be a bit more intuitive is to make the `server` case on `Source` have an associated value since `cache` sources will always be complete. The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received. This solution becomes invalid though once the cache can return partial responses, with that in mind maybe option 1 is better.
```

Comment:
Interesting thought!  But I agree for the reason you stated that we should stick with option 1.